### ISSUE TITLE
org.bouncycastle.cms.CMSException: Could not create DES cipher

### ISSUE THREAD

**Issue by pauldavidgilligan-msm** [jscep/jscep#30]
Intermittent Exception :  Using BC 1.51 + latest JCEP
```
[dev-ops@localhost ife-toolbelt]$ /opt/jdk1.7.0_71/bin/java -version
java version "1.7.0_71"
Java(TM) SE Runtime Environment (build 1.7.0_71-b14)
Java HotSpot(TM) 64-Bit Server VM (build 24.71-b01, mixed mode)
```

EJBCA is running on OpenJDK in Docker container which might be a problem.

```
org.jscep.transaction.TransactionException: org.jscep.message.MessageDecodingException: org.bouncycastle.cms.CMSException: Could not create DES cipher
	at org.jscep.transaction.EnrollmentTransaction.send(EnrollmentTransaction.java:123)
	at org.jscep.client.Client.send(Client.java:680)
	at org.jscep.client.Client.enrol(Client.java:635)
	at internal.msm.tools.ife.scep.client.AutoEnroller.getCertificate(AutoEnroller.java:150)
	at internal.msm.tools.ife.spring.shell.commands.SCEPCommands.enroll(SCEPCommands.java:134)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.springframework.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:202)
	at org.springframework.shell.core.SimpleExecutionStrategy.invoke(SimpleExecutionStrategy.java:64)
	at org.springframework.shell.core.SimpleExecutionStrategy.execute(SimpleExecutionStrategy.java:57)
	at org.springframework.shell.core.AbstractShell.executeCommand(AbstractShell.java:127)
	at org.springframework.shell.core.JLineShell.promptLoop(JLineShell.java:533)
	at org.springframework.shell.core.JLineShell.run(JLineShell.java:179)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.jscep.message.MessageDecodingException: org.bouncycastle.cms.CMSException: Could not create DES cipher
	at org.jscep.message.PkcsPkiEnvelopeDecoder.decode(PkcsPkiEnvelopeDecoder.java:95)
	at org.jscep.message.PkiMessageDecoder.decode(PkiMessageDecoder.java:163)
	at org.jscep.transaction.Transaction.decode(Transaction.java:107)
	at org.jscep.transaction.EnrollmentTransaction.send(EnrollmentTransaction.java:121)
	... 15 more
Caused by: org.bouncycastle.cms.CMSException: Could not create DES cipher
	at org.jscep.message.PkcsPkiEnvelopeDecoder$InternalKeyTransEnvelopedRecipient.getRecipientOperator(PkcsPkiEnvelopeDecoder.java:134)
	at org.bouncycastle.cms.KeyTransRecipientInformation.getRecipientOperator(Unknown Source)
	at org.bouncycastle.cms.RecipientInformation.getContentStream(Unknown Source)
	at org.bouncycastle.cms.RecipientInformation.getContent(Unknown Source)
	at org.jscep.message.PkcsPkiEnvelopeDecoder.decode(PkcsPkiEnvelopeDecoder.java:91)
	... 18 more
Caused by: java.security.InvalidKeyException: Unwrapping failed
	at com.sun.crypto.provider.RSACipher.engineUnwrap(RSACipher.java:447)
	at javax.crypto.Cipher.unwrap(Cipher.java:2466)
	at org.jscep.message.PkcsPkiEnvelopeDecoder$InternalKeyTransEnvelopedRecipient.unwrapKey(PkcsPkiEnvelopeDecoder.java:155)
	at org.jscep.message.PkcsPkiEnvelopeDecoder$InternalKeyTransEnvelopedRecipient.getRecipientOperator(PkcsPkiEnvelopeDecoder.java:130)
	... 22 more
Caused by: javax.crypto.BadPaddingException: Decryption error
	at sun.security.rsa.RSAPadding.unpadV15(RSAPadding.java:380)
	at sun.security.rsa.RSAPadding.unpad(RSAPadding.java:291)
	at com.sun.crypto.provider.RSACipher.doFinal(RSACipher.java:365)
	at com.sun.crypto.provider.RSACipher.engineUnwrap(RSACipher.java:442)
	... 25 more
```

Here is the cert from EJBCA:

```
14:45:13,080 INFO  [org.cesecore.audit.impl.log4j.Log4jDevice] (http--0.0.0.0-8080-1) 2015-01-07 14:45:13+00:00;CERT_CREATION;SUCCESS;CERTIFICATE;CORE;ScepServlet: 172.17.42.1;-1024587024;3D5244C2FAF87D38;*.web.be.msm.internal;subjectdn=CN=*.web.be.msm.internal,OU=OPS,O=Moneysupermarket.com Group PLC,L=Chester,ST=Cheshire,C=UK;certprofile=9;issuancerevocationreason=-1;cert=MIIF0zCCA7ugAwIBAgIIPVJEwvr4fTgwDQYJKoZIhvcNAQELBQAwUjEaMBgGA1UEAwwRcm9vdC5tc20uaW50ZXJuYWwxJzAlBgNVBAoMHk1vbmV5c3VwZXJtYXJrZXQuY29tIEdyb3VwIFBMQzELMAkGA1UEBhMCVUswHhcNMTUwMTA3MTQzNTEzWhcNMTcwMTA2MTQzNTEzWjCBiTEeMBwGA1UEAwwVKi53ZWIuYmUubXNtLmludGVybmFsMQwwCgYDVQQLDANPUFMxJzAlBgNVBAoMHk1vbmV5c3VwZXJtYXJrZXQuY29tIEdyb3VwIFBMQzEQMA4GA1UEBwwHQ2hlc3RlcjERMA8GA1UECAwIQ2hlc2hpcmUxCzAJBgNVBAYTAlVLMIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEAm9RWVSRENy6bCWcb7RQfwkNw97zFoNY+H+CLZ8Xn0m9B0vj8RtVgyyZxH4yZmyVvPA7EknZ9/beVjfBCuIMafd0MxA+Mv1ugEOoah0VdspsSzFQGCKLecBhZ2wURS0DEppN01vOM7584eLNdBwSEo23r3vgqoIbApc04grW1svuwioLN+6vnZNX5nlwwm7HnqAlF1f7ML/I3D73ICnPKx8KLYOjmvCifDqIBaz3g3yccHz4Kql6WPNTE6VJJewIHyqpfzq/Sest4N9de8WdTwlrcBGgQqE5vdbIOt5B144m6476eyI39iAVg+D10MNG5FfQPsza3Ynji+7SyZQ/ls3vKO8Zn40TJxK4azl3a9R/5s+WpgrgLjuPFRl2sahei9pDpRzbRYyychEqR6y+M/LEl3xazL0ZIKzksFkRi/ZJC+onqi+IEOOabPDABj7pxEgrVFuNfrTaYmTd/JvNlgBeitVqQUbxZ9q7dN5vCUWQcc8xCsg1vJFWF4I0XQN2GtJQfmzADQpzYmX9ghY9tQHiH/t8KGflqSYcotd3l8NNPy2cQdMx6oLBJgZoIhJhXS9VLC8zoMUhDFTPmNf7uPL4U4ME5NWkR44P4IpcBC6P5At3SXEMRFWgb61jBApC6kQiL+5hj/vgIIGYM8Mw9wwAzpsTYP2XpC9P8O/pRKG0CAwEAAaN1MHMwHQYDVR0OBBYEFK4DckfIB1Q6CjeDNHf4b/ncJNqYMAwGA1UdEwEB/wQCMAAwHwYDVR0jBBgwFoAUAPS9MVxiV4LZ5ERB3d6ZWugWUuswDgYDVR0PAQH/BAQDAgWgMBMGA1UdJQQMMAoGCCsGAQUFBwMBMA0GCSqGSIb3DQEBCwUAA4ICAQAhE0xN86qj2lDaB6eGc/qDJvIsa1pgKt/NbPMTlwsIIlw4sMY+6J43CgW7nzYcHvE4h4n/Zvzm59IzBGQ9CyIcWyDUz1YN58tXoclo56NeGnYJHuArGAvqnPcp4TnNbTwCWVsqbVe3Nw26XRG4OAuO6ld31JMdZW/ArPseZx+Zr0A0X2mccx2H/6qR3AnTr+7WI1rlqUCSE9ElFI5ylNuVxnvwE/ai8dIRiLEI74krc+U7qh0Kbf82DDOwBvp1YBRCQw+HsD1tcp+6lRwuDMr+orWmx8MHmKqIO9PG8YBgPF3GsWTNT5cXSxe0R/gteWHdbDjkJOLKCl1ZL719KYdeexDHnLmKNbfWG49CHSHRN6taVb8tXLo9+r9HFGxWYu31uI0N9Wg1oUVY/QwHcHmJGjRdR9/m5TbEupXykfWrU9wJOG7rzvuqxuUZJtCMnJoBjWnOZDKcQ3S3I9F3LQNZ9XTbFMcodykjibYeBoMDsos48khg0CngSoeocitPvMTOTRuBzGoxHaB0UID0uFF/cMPhCFHzSfsVTD8jkH6IH6ixtzB1R/xUzbq1R1RkHYH7aL5JZviE57z5ajll7e+c59wTyY/L+VRcHoY4M4D5muNc81yZDtW1ErbEyuhwJiwMPWM0eyszK9s2t2qVYznclof4xubNdta5V0qEuiDDDg==
```

**Comment by seize-the-dave** [comment]
@pauldavidgilligan-msm can you try with BC 1.50?

**Comment by pauldavidgilligan-msm** [comment]
Checked the X.509 cert from EJBCA looks great so not sure why JSCEP is having an issue decoding.

**Comment by pauldavidgilligan-msm** [comment]
No luck with 1.50 I generate the RSA key via:
```
        final SecureRandom sr1 = SecureRandom.getInstance("SHA1PRNG");
        final SecureRandom sr2 = SecureRandom.getInstance("SHA1PRNG");

        final KeyPairGenerator keyPairGenerator = KeyPairGenerator.getInstance("RSA", "BC");
        keyPairGenerator.initialize(this.size, sr1);

        final KeyPair keyPair = keyPairGenerator.genKeyPair();

        final PublicKey pubkey = keyPair.getPublic();
        LOGGER.debug("Saving public key to [" + pubkey.getFormat() + "] " + pubpath);
        byte[] publicKeyBytes = pubkey.getEncoded();
        try (FileOutputStream fos = new FileOutputStream(pubpath.toFile())) {
            fos.write(publicKeyBytes);
        }

        final PrivateKey prtvkey = keyPair.getPrivate();
        final JceOpenSSLPKCS8EncryptorBuilder encryptorBuilder = new JceOpenSSLPKCS8EncryptorBuilder(PKCS8Generator.PBE_SHA1_3DES);
        encryptorBuilder.setIterationCount(10000);
        encryptorBuilder.setRandom(sr2);
        encryptorBuilder.setPasssword(password.toCharArray());
        OutputEncryptor oe = encryptorBuilder.build();
        PKCS8Generator gen = new JcaPKCS8Generator(prtvkey, oe);

        LOGGER.debug("Saving private key to [PKCS8]" + pripath_pem);
        try (PemWriter writer = new PemWriter(new PrintWriter(new FileOutputStream(pripath_pem.toFile())))) {
            writer.writeObject(gen);
        }
        return keyPair;

```

**Comment by pauldavidgilligan-msm** [comment]
RSA key validates so I don't think as issue.

**Comment by seize-the-dave** [comment]
@pauldavidgilligan-msm could you rerun with Log4J logging set to DEBUG or TRACE for org.jscep please?

**Comment by pauldavidgilligan-msm** [comment]
Strange... added in log4J and debug now working again project is https://github.com/pauldavidgilligan-msm/ife-toolbelt

Now I get:
```
Got CA certificate from configuration file [/etc/ife-toolbelt.cfg]
spring> ife scep enroll --CN n1.web.be.msm.internal
2015-01-07 16:06:45 WARN  BCPEMWriter:90 - Wildcard CN match, overriding n1.web.be.msm.internal with *.web.be.msm.internal
2015-01-07 16:06:45 DEBUG BCPEMReader:88 - Reading X509 PEM from /etc/pki/msm/wildcard.web.be.msm.internal.crt
2015-01-07 16:06:45 DEBUG BCPEMReader:93 - Loaded X509 PEM DN: CN=*.web.be.msm.internal,OU=OPS,O=Moneysupermarket.com Group PLC,L=Chester,ST=Cheshire,C=UK
2015-01-07 16:06:45 DEBUG BCPEMReader:60 - Reading RSA PEM from /etc/pki/msm/wildcard.web.be.msm.internal.private.pem
2015-01-07 16:06:45 DEBUG BCPEMReader:64 - JcaPEMKeyConverter loaded
2015-01-07 16:06:45 DEBUG BCPEMReader:72 - Decrypting PKCS8 PEM file
2015-01-07 16:06:45 DEBUG BCPEMReader:106 - Reading CSR PEM from /etc/pki/msm/wildcard.web.be.msm.internal.csr
2015-01-07 16:06:45 DEBUG BCPEMReader:111 - Loaded CSR PEM DN: CN=*.web.be.msm.internal,OU=OPS,O=Moneysupermarket.com Group PLC,L=Chester,ST=Cheshire,C=UK
2015-01-07 16:06:45 DEBUG Client:604 - Enrolling certificate with CA
2015-01-07 16:06:45 DEBUG Client:607 - Certificate is self-signed
2015-01-07 16:06:45 DEBUG Client:218 - Determining capabilities of SCEP server
2015-01-07 16:06:45 DEBUG Client:268 - Retrieving current CA certificate
2015-01-07 16:06:45 DEBUG Client:735 - Requesting certificate verification.
2015-01-07 16:06:45 DEBUG Client:750 - Certificate verification passed.
2015-01-07 16:06:45 DEBUG Client:290 - Verifying signature of RA certificate
2015-01-07 16:06:45 DEBUG Client:292 - RA and CA are identical
2015-01-07 16:06:45 DEBUG Client:290 - Verifying signature of RA certificate
2015-01-07 16:06:45 DEBUG Client:292 - RA and CA are identical
2015-01-07 16:06:45 DEBUG Client:218 - Determining capabilities of SCEP server
2015-01-07 16:06:45 DEBUG Client:268 - Retrieving current CA certificate
2015-01-07 16:06:45 DEBUG Client:735 - Requesting certificate verification.
2015-01-07 16:06:45 DEBUG Client:750 - Certificate verification passed.
2015-01-07 16:06:45 DEBUG Client:290 - Verifying signature of RA certificate
2015-01-07 16:06:45 DEBUG Client:292 - RA and CA are identical
2015-01-07 16:06:45 DEBUG Client:290 - Verifying signature of RA certificate
2015-01-07 16:06:45 DEBUG Client:292 - RA and CA are identical
2015-01-07 16:06:45 DEBUG Client:218 - Determining capabilities of SCEP server
2015-01-07 16:06:45 DEBUG Client:629 - SHA-1 PKCS#10 Fingerprint: [cc857e10930117050a737066fd99661607554646]
2015-01-07 16:06:46 WARN  PkiMessageDecoder:126 - Unable to verify message because the signedData contained no certificates.
2015-01-07 16:06:46 INFO  AutoEnroller:152 - Enrolled with CA # 202887761846002645
2015-01-07 16:06:46 INFO  AutoEnroller:160 - Enrolled C=UK, ST=Cheshire, L=Chester, O=Moneysupermarket.com Group PLC, OU=OPS, CN=*.web.be.msm.internal
2015-01-07 16:06:46 INFO  AutoCompute:147 - Saving enrolled cert to /etc/pki/msm/wildcard.web.be.msm.internal.crt with CN *.web.be.msm.internal
2015-01-07 16:06:46 DEBUG AbstractGenerator:92 - /etc/pki/msm is valid.
2015-01-07 16:06:46 DEBUG BCSystem:38 - java.home set /opt/jdk1.7.0_71/jre
2015-01-07 16:06:46 DEBUG KeyStoreGenerator:119 - Writing keystore to /etc/pki/msm/wildcard.web.be.msm.internal.keystore

Enrolled with CA from configuration file [/etc/ife-toolbelt.cfg] for CN [n1.web.be.msm.internal]
```

**Comment by pauldavidgilligan-msm** [comment]
I always got 2015-01-07 16:06:46 WARN  PkiMessageDecoder:126 - Unable to verify message because the signedData contained no certificates. 

though...

**Comment by pauldavidgilligan-msm** [comment]
```
2015-01-07 16:14:12 DEBUG Client:629 - SHA-1 PKCS#10 Fingerprint: [cc857e10930117050a737066fd99661607554646]
2015-01-07 16:14:12 WARN  PkiMessageDecoder:126 - Unable to verify message because the signedData contained no certificates.
2015-01-07 16:14:12 ERROR AutoEnroller:172 - *.web.be.msm.internal
org.jscep.transaction.TransactionException: org.jscep.message.MessageDecodingException: org.bouncycastle.cms.CMSException: Could not create DES cipher
	at org.jscep.transaction.EnrollmentTransaction.send(EnrollmentTransaction.java:123)
	at org.jscep.client.Client.send(Client.java:680)
	at org.jscep.client.Client.enrol(Client.java:635)
	at internal.msm.tools.ife.scep.client.AutoEnroller.getCertificate(AutoEnroller.java:150)
	at internal.msm.tools.ife.spring.shell.commands.SCEPCommands.enroll(SCEPCommands.java:134)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.springframework.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:202)
	at org.springframework.shell.core.SimpleExecutionStrategy.invoke(SimpleExecutionStrategy.java:64)
	at org.springframework.shell.core.SimpleExecutionStrategy.execute(SimpleExecutionStrategy.java:57)
	at org.springframework.shell.core.AbstractShell.executeCommand(AbstractShell.java:127)
	at org.springframework.shell.core.JLineShell.promptLoop(JLineShell.java:533)
	at org.springframework.shell.core.JLineShell.run(JLineShell.java:179)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.jscep.message.MessageDecodingException: org.bouncycastle.cms.CMSException: Could not create DES cipher
	at org.jscep.message.PkcsPkiEnvelopeDecoder.decode(PkcsPkiEnvelopeDecoder.java:95)
	at org.jscep.message.PkiMessageDecoder.decode(PkiMessageDecoder.java:163)
	at org.jscep.transaction.Transaction.decode(Transaction.java:107)
	at org.jscep.transaction.EnrollmentTransaction.send(EnrollmentTransaction.java:121)
	... 15 more
Caused by: org.bouncycastle.cms.CMSException: Could not create DES cipher
	at org.jscep.message.PkcsPkiEnvelopeDecoder$InternalKeyTransEnvelopedRecipient.getRecipientOperator(PkcsPkiEnvelopeDecoder.java:134)
	at org.bouncycastle.cms.KeyTransRecipientInformation.getRecipientOperator(Unknown Source)
	at org.bouncycastle.cms.RecipientInformation.getContentStream(Unknown Source)
	at org.bouncycastle.cms.RecipientInformation.getContent(Unknown Source)
	at org.jscep.message.PkcsPkiEnvelopeDecoder.decode(PkcsPkiEnvelopeDecoder.java:91)
	... 18 more
Caused by: java.security.InvalidKeyException: Unwrapping failed
	at com.sun.crypto.provider.RSACipher.engineUnwrap(RSACipher.java:447)
	at javax.crypto.Cipher.unwrap(Cipher.java:2466)
	at org.jscep.message.PkcsPkiEnvelopeDecoder$InternalKeyTransEnvelopedRecipient.unwrapKey(PkcsPkiEnvelopeDecoder.java:155)
	at org.jscep.message.PkcsPkiEnvelopeDecoder$InternalKeyTransEnvelopedRecipient.getRecipientOperator(PkcsPkiEnvelopeDecoder.java:130)
	... 22 more
Caused by: javax.crypto.BadPaddingException: Decryption error
	at sun.security.rsa.RSAPadding.unpadV15(RSAPadding.java:380)
	at sun.security.rsa.RSAPadding.unpad(RSAPadding.java:291)
	at com.sun.crypto.provider.RSACipher.doFinal(RSACipher.java:365)
	at com.sun.crypto.provider.RSACipher.engineUnwrap(RSACipher.java:442)
	... 25 more

```

**Comment by seize-the-dave** [comment]
jscep keeps no state itself so I would expect the success to be down to EJBCA.

Your logs appear to be missing everything below the org.jscep category, e.g. org.jscep.message.  Have you got a custom Log4J configuration?

**Comment by pauldavidgilligan-msm** [comment]
yes what would be the best log4j.properties setup to debug please ?

**Comment by seize-the-dave** [comment]
You can copy the following:

https://github.com/jscep/jscep/blob/master/src/test/resources/log4j.properties

**Comment by pauldavidgilligan-msm** [comment]
ah that is better:

2015-01-07 16:21:35 DEBUG PkiMessageDecoder:81 - Decoding pkiMessage
2015-01-07 16:21:35 DEBUG PkiMessageDecoder:81 - Decoding pkiMessage
2015-01-07 16:21:35 DEBUG PkiMessageDecoder:207 - pkiMessage version: 1
2015-01-07 16:21:35 DEBUG PkiMessageDecoder:207 - pkiMessage version: 1
2015-01-07 16:21:35 DEBUG PkiMessageDecoder:208 - pkiMessage contentInfo contentType: 1.2.840.113549.1.7.1
2015-01-07 16:21:35 DEBUG PkiMessageDecoder:208 - pkiMessage contentInfo contentType: 1.2.840.113549.1.7.1
2015-01-07 16:21:35 DEBUG PkiMessageDecoder:94 - pkiMessage digest algorithm: 1.2.840.113549.2.5
2015-01-07 16:21:35 DEBUG PkiMessageDecoder:94 - pkiMessage digest algorithm: 1.2.840.113549.2.5
2015-01-07 16:21:35 DEBUG PkiMessageDecoder:96 - pkiMessage encryption algorithm: 1.2.840.113549.1.1.1
2015-01-07 16:21:35 DEBUG PkiMessageDecoder:96 - pkiMessage encryption algorithm: 1.2.840.113549.1.1.1
2015-01-07 16:21:35 WARN  PkiMessageDecoder:126 - Unable to verify message because the signedData contained no certificates.
2015-01-07 16:21:35 WARN  PkiMessageDecoder:126 - Unable to verify message because the signedData contained no certificates.
2015-01-07 16:21:35 DEBUG PkiMessageDecoder:133 - pkiMessage has 8 signed attributes:
2015-01-07 16:21:35 DEBUG PkiMessageDecoder:133 - pkiMessage has 8 signed attributes:

Command failed java.lang.ClassCastException: org.bouncycastle.asn1.ASN1ObjectIdentifier cannot be cast to org.bouncycastle.asn1.DERObjectIdentifier

**Comment by seize-the-dave** [comment]
This problem is only present when logging is on (and should work in BC 1.50).  I noticed it when I created a branch for testing 1.51, and it's fixed here:

https://github.com/jscep/jscep/commit/615b241a5a8778783e206a83b2cb977f1a8a88ce

If you can build jscep from source, this problem should go away.  I'll be able to do a new release this evening (GMT), but if you can try it out, you'll be able to move beyond this.

**Comment by pauldavidgilligan-msm** [comment]
nice thanks for the quick reply.

**Comment by seize-the-dave** [comment]
@pauldavidgilligan-msm 

I've just released 2.2.0, which contains the fix linked in the [comment above](https://github.com/jscep/jscep/issues/30#issuecomment-69047246), any also upgrades the BC dependency to v1.51.

**Comment by pauldavidgilligan-msm** [comment]
Thanks looking much more stable so I will close this.
